lib/vu_state.go

```diff
@@ -93,65 +93,39 @@ type State struct {
 	GetScenarioGlobalVUIter func() uint64
 }
 
-// TagMap is a safe-concurrent Tags lookup.
-type TagMap struct {
+// VUStateTags wraps the current VU's tags and ensures a thread-safe way to
+// access and modify them exists. This is necessary because the VU tags and
+// metadata can be modified from the JS scripts via the `vu.tags` API in the
+// `k6/execution` built-in module.
+type VUStateTags struct {
 	mutex sync.RWMutex
 	tags  *metrics.TagSet
+	// TODO: Add metadata map[string]string
 }
 
-// NewTagMap creates a TagMap,
-// if a not-nil map is passed then it will be used as the internal map
-// otherwise a new one will be created.
-func NewTagMap(tags *metrics.TagSet) *TagMap {
+// NewVUStateTags initializes a new VUStateTags and returns it. It's important
+// that tags is not nil and initialized via metrics.Registry.RootTagSet().
+func NewVUStateTags(tags *metrics.TagSet) *VUStateTags {
 	if tags == nil {
-		panic("the metrics.TagSet must be initialized for creating a new lib.TagMap")
+		panic("the metrics.TagSet must be initialized for creating a new lib.VUStateTags")
 	}
-	return &TagMap{
-		tags:  tags,
+	return &VUStateTags{
 		mutex: sync.RWMutex{},
+		tags:  tags,
+		// metadata is intentionally nil by default
 	}
 }
 
-// BranchOut creates a TagSet starting from the state of the current Set.
-func (tg *TagMap) BranchOut() *metrics.TagSet {
+// GetCurrentValues returns the value of the VU tags in a thread-safe way.
+func (tg *VUStateTags) GetCurrentValues() *metrics.TagSet {
 	tg.mutex.RLock()
 	defer tg.mutex.RUnlock()
-	return tg.tags.BranchOut()
+	return tg.tags
 }
 
-// Set sets a Tag.
-func (tg *TagMap) Set(k, v string) {
+// Modify allows the thread-safe modification of the current VU tags.
+func (tg *VUStateTags) Modify(callback func(currentTags *metrics.TagSet) (newTags *metrics.TagSet)) {
 	tg.mutex.Lock()
 	defer tg.mutex.Unlock()
-	tg.tags.AddTag(k, v)
-}
-
-// Get returns the Tag value and true
-// if the provided key has been found.
-func (tg *TagMap) Get(k string) (string, bool) {
-	tg.mutex.RLock()
-	defer tg.mutex.RUnlock()
-	return tg.tags.Get(k)
-}
-
-// Len returns the number of the set keys.
-func (tg *TagMap) Len() int {
-	tg.mutex.RLock()
-	defer tg.mutex.RUnlock()
-	return tg.tags.Len()
-}
-
-// Delete deletes a map's item based on the provided key.
-func (tg *TagMap) Delete(k string) {
-	tg.mutex.Lock()
-	defer tg.mutex.Unlock()
-	tg.tags.Delete(k)
-}
-
-// Clone returns a map with the entire set of items.
-func (tg *TagMap) Clone() map[string]string {
-	tg.mutex.RLock()
-	defer tg.mutex.RUnlock()
-
-	return tg.tags.Map()
+	tg.tags = callback(tg.tags)
```

Comment:
Not certain this is better than having `Set` and `Delete` The only place where this is used not for this (in the vu Activate) the lib.state should not be used by anything. So it will be safe to just not do it like this ;)

Reply:
`Set()` and `Delete()` and getting the current values are not atomic, we might have some weird indeterminate states with them, even some lost tags that get overwritten... This locks the `State.Tags` for the whole modification operation - you get the current state and emit the new state, no chance of a race.

Reply:
Thats all nice and true ... but where does this matter? AFAIK all the cases we have are 
1. we have to set 1 thing irregardless of anything in the map
2. we have to get 1 thing irregardless of anything in the map
3. the initialization of this ... which doesn't need any of this as it's before there can be any race conditions. 

So for me this just forces you to use lambdas .. .and I know how much you love lambdas but that isn't a reason :P

Reply:
There are a few reasons why I don't want to change this:
- it might not matter anywhere _now_, but it might matter in the future, I don't know... this lambda is so easy to use that I don't consider it a problem worth changing and potentially shooting ourselves in the foot later
- it's more inline with the immutable nature of the atlas Nodes and `metrics.TagSet` - you disliked the mutable `TagSet` that was originally part of https://github.com/grafana/k6/pull/2594, why do you want a mutable
- this paradigm works much better with the unindexable tags, where through `k6/execution.vu.tags` a user actually might actually transform an indexed tag (stored in atlas) into an unindexed one (stored in the string metadata map), this probably should be atomic to avoid any issues